cmds/vbe.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * VESA VBE modesetting
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Adam Greloch
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <syspage.h>
+
+
+#define PTR_TO_SEG(ptr)       ((addr_t)ptr >> 4)
+#define SEGOFS_TO_PTR(segofs) ((addr_t)((segofs & 0xffff0000) >> 12) + (segofs & 0xffff))
+
+#define SUPPORTS_LFB (0x80)
+
+
+typedef struct {
+	u16 width;
+	u16 height;
+	u16 bpp;
+} video_mode_t;
+
+
+typedef struct {
+	char signature[4];
+	u16 version;
+	u32 oem;
+	u32 capabilities;
+	u32 video_modes;
+	u16 video_memory;
+	u16 software_rev;
+	u32 vendor;
+	u32 product_name;
+	u32 product_rev;
+	char reserved[222];
+	char oem_data[256];
+} __attribute__((packed)) vbe_info_t;
+
+
+typedef struct {
+	u16 attributes;
+	u8 window_a;
+	u8 window_b;
+	u16 granularity;
+	u16 window_size;
+	u16 segment_a;
+	u16 segment_b;
+	u32 win_func_ptr;
+	u16 pitch;
+	u16 width;
+	u16 height;
+	u8 w_char;
+	u8 y_char;
+	u8 planes;
+	u8 bpp;
+	u8 banks;
+	u8 memory_model;
+	u8 bank_size;
+	u8 image_pages;
+	u8 reserved0;
+
+	u8 red_mask;
+	u8 red_position;
+	u8 green_mask;
+	u8 green_position;
+	u8 blue_mask;
+	u8 blue_position;
+	u8 reserved_mask;
+	u8 reserved_position;
+	u8 direct_color_attributes;
+
+	u32 framebuffer;
+	u32 off_screen_mem_off;
+	u16 off_screen_mem_size;
+	u8 reserved1[206];
+} __attribute__((packed)) vbe_mode_info_t;
+
+
+static video_mode_t preferred_mode = { 1280, 800, 32 };
+static video_mode_t fallback_mode = { 1024, 768, 32 };
+
+
+/* ADDR_VBE_INFO, ADDR_VBE_MODE_INFO should be aligned to 0x10, so that they can
+ * be referenced from 8086 emulation mode using segment register only (es=(ADDR >> 4), di=0) */
+static vbe_info_t *vbe_info = (vbe_info_t *)ADDR_VBE_INFO;
+static vbe_mode_info_t *vbe_mode_info = (vbe_mode_info_t *)ADDR_VBE_MODE_INFO;
+
+
+static int _get_vbe_info(void)
+{
+	/* indicate support for VBE 2.0+ */
+	hal_memset(vbe_info, 0, sizeof(vbe_info_t));
+	hal_strcpy(vbe_info->signature, "VBE2");
+
+	int ax = 0x4f00;
+
+	__asm__ volatile(
+		"pushl $0x10; "
+		"pushl $0x0; "
+		"pushl %1; "
+		"xorw %%di, %%di; "
+		"call _interrupts_bios; "
```

Comment:
Apart from calling `_interrupt_bios` I also need to have the DI register zeroed before the call and pass the VBE function number in the AX. Not sure if performing a normal call would allow this

Reply:
I can't find _interrupt_bios implementation, so it's hard to me to suggest anything

Reply:
It's in `plo/hal/ia32/_interrupts.S`. This function passes i.e. AX, DI registers to the performed interrupt. The DI denotes the offset of ES:DI segment:offset address that is passed to the interrupt done in 8086 mode and remains untouched by the `_interrupt_bios` function. Setting it to zero is crucial for making sure the address argument is correctly passed to the bios function. A similar asm code to mine is present in `plo/devices/disk-bios/disk-bios.c`, e.g. L134

Reply:
Ok, some option is to pass DI as an argument, but even if it's good idea, it's out of scope of this PR. So it's ok as is